### DISCUSSION TITLE
Periodically Update YARP destinations

### DISCUSSION THREAD

**Discussion by eyalg-msft** [dotnet/yarp#1963] [Q&A] [answered]
Background:
We have migrates to YARP from IIS + ARR on top of VMs. YARP forwards the requests to our application backend VMs.
We need a dynamic update (periodic) of our destination list. Is there any built in way to provide YARP with an update delegate which returns (routes, clusters), or even just clusters per route, and then YARP will update the clusters if needed?

Since we didn't find anything like that, I want to share what we did, 
I would be happy to hear your thoughts of this solution. Is this the right way to do it? do you have a recommendation to do it differently?

So here goes:

1.Added a BackgroundService which I called "ConfigurationUpdaterBackgroundService" that runs every 3 minutes.

2.On the ConfigureServices when creating our new BackgroundService we use the service provider to get the IProxyConfigProvider and pass it to the background service constructor.

services.AddHostedService<ConfigurationUpdaterBackgroundService>(serviceProvider => new ConfigurationUpdaterBackgroundService(proxyBuilder, ServiceInstanceFinder, Configuration, serviceProvider.GetRequiredService<IProxyConfigProvider>()));

3.The BackgroundService will now periodically check for updates in the list (we have another service to get it), and if needed it wil call the Update function of YARP:

this.reverseProxyInMemoryConfigProvider.Update(routes, clusters);

**Comment by Tratcher** [comment, marked as answer]
See https://microsoft.github.io/reverse-proxy/articles/config-providers.html#reload

Yes, that's a common pattern. IProxyConfigProvider is the abstraction for providing and reloading config. Using the existing InMemoryConfigProvider is a fine way to plug in when you're pulling your config from somewhere else. Are you using the new public one from the 2.0 previews?

**Reply by eyalg-msft** [reply]
Thanks, Chris, for the answer.
I am using version 1.1 the InMemoryConfigProvider.Update is public AFAIU. 

In my solution I have to cast the IProxyConfigProvider to the InMemoryConfigProvider to access the Update method.
Is it possible that there are more than one IProxyConfigProvider? Can I get specifically the InMemoryConfigProvider?

Did you think of exposing some API for periodic updates or any evented updates?
Maybe even in a configuration way, for example in my case give the amount of time to check if we need to update the configuration list. I can only guess that in real world scenarios the destination list is dynamic and needs to be updated when new destinations are added. I guess this could really simplify things if it was just a flag to turn on to revalidate the list periodically. What do you think?

**Reply by Tratcher** [reply]
InMemoryConfigProvider wasn't included publicly in the dll until 2.0 previews. Before that many samples had a copy of it. 

When we included it in 2.0 we added a pattern where you could get the InMemoryConfigProvider directly so you could update it.
https://github.com/microsoft/reverse-proxy/blob/65af2875bb552402575d49a225937ae5fce274e4/samples/ReverseProxy.Code.Sample/Startup.cs#L44

We haven't explored periodic updates since that can be built on top of the current model. Many environments have their own change signal that can be used instead of a timer, and a change signal can be much more efficient. Where are you reading your config from?

**Reply by eyalg-msft** [reply]
This is exactly what I was missing. I am getting it by interface IProxyConfigProvider, then I am forced to cast it to InMemoryConfigProvider to use the update method. I will add a task to update it once we move to YARP 2.0.

Regarding the service destinations discovery. This is internal Microsoft WAC component. It's how we discover the service instances on WAC DC using EAP. Unfortunately, it's not event based, we need to periodically query for updates.
